### PR TITLE
Feature/apl 527 implement sharding

### PR DESCRIPTION
**Pull request by yuriy-larin-firstbridge**
Genesis loading fixes

### CODE REVIEW

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/BlockchainProcessorImpl.java [hunk 1]
```diff
@@ -626,16 +627,16 @@ private void continuedDownloadOrTryImportGenesisShard() {
...
-        setGetMoreBlocks(false); // turn off automatic blockchain downloading
+        // NEW START-UP logic, try import genesis OR start downloading shard zip data
+        suspendBlockchainDownloading(); // turn off automatic blockchain downloading
         log.warn("NODE IS WAITING FOR no/shard decision and proceeding with necessary data by ShardPresentEventType....");
 
-//        FileDownloader downloader; ???
-//        downloader.startDownload(id); ???
-//        FileDownloadDecision decision = downloader.prepareForDownloading(); ???
-//        FileDownloadInfo fdi = downloader.getDownloadInfo(); ??
+        ShardDownloader shardDownloader = CDI.current().select(ShardDownloader.class).get();
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/BlockchainProcessorImpl.java]: Make it injectable or move to separate component

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java [hunk 2]
```diff
@@ -688,7 +688,7 @@ protected int update(int nextUpdate) {
             case 252:
                 // SHARDING meta-info inside main database
                 apply("CREATE TABLE IF NOT EXISTS shard (shard_id BIGINT NOT NULL, shard_hash VARBINARY, " +
-                        "shard_height INT not null default 0, shard_state BIGINT default 0, zip_hash_crc VARBINARY, generator_ids ARRAY )");
+                        "shard_height INT not null default 0, shard_state BIGINT default 0, zip_hash_crc VARBINARY, generator_ids ARRAY NOT NULL)");
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: Behavior was changed, we should allow null here to save shard record during sharding process multiple times. ALready fixed in develop

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/get/GetBlockchainStatus.java [hunk 3]
```diff
@@ -53,9 +53,15 @@ public JSONObject processRequest(HttpServletRequest req) {
         response.put("version", Constants.VERSION.toString());
         response.put("time", timeService.getEpochTime());
         Block lastBlock = lookupBlockchain().getLastBlock();
-        response.put("lastBlock", lastBlock.getStringId());
-        response.put("cumulativeDifficulty", lastBlock.getCumulativeDifficulty().toString());
-        response.put("numberOfBlocks", lastBlock.getHeight() + 1);
+        if (lastBlock != null) { // TODO: YL I hope that is temporary decision to prevent NPE
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/get/GetBlockchainStatus.java]: Better throw exception

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/BlockchainProcessorImpl.java (hunk 1 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/BlockchainProcessorImpl.java]: Done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java (hunk 2 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: Reverted back

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/get/GetBlockchainStatus.java (hunk 3 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/get/GetBlockchainStatus.java]: That's for UI only... during starting up node while no genesis/shard snapshot imported

**Pull request merged**